File: gifts.gl2/gl_tutor/src/workshop/menu8.c
/*  menu8.c -- This program allows the user a pop-up menu	*/
/*  interface to an application program.  In this example,	*/
/*  the RIGHT MOUSE button controls a popup menu.  This popup	*/
/*  menu can be used to exit the program.  Intention to exit	*/
/*  the program must be reconfirmed by the user responding to	*/
/*  a second popup menu.  Popup menu support is ONLY available	*/
/*  in the window manager.					*/
/*  graphics library calls introduced:  defpup(), newpup(),	*/
/*		addtopup(), dopup().				*/

#include "gl.h"
#include "device.h"

/*  new global variables for menus		*/

int     menu,
        killmenu;

/*  the main() routine has not changed since the last example.	*/

main () {
    initialize ();
    drawimage ();
    while (TRUE) {	/*  loop forever.  swapbuffers and process	*/
	while (!qtest())/*  input while looping				*/
	    swapbuffers();	
	processinput();
    }
}

/*  In initialize(), queue the RIGHTMOUSE button for popup menus.	*/
/*  Also the pup calls are used to create popup menus.			*/

initialize () {
    int i;

    prefsize (450, 450);
    winopen ("diamond");

    doublebuffer();	/* display mode to become double buffer */
    gconfig();		/* display mode change takes effect	*/

    qdevice(LEFTMOUSE); /* queue the LEFTMOUSE device		*/
    qdevice(RIGHTMOUSE);/* queue the RIGHTMOUSE device		*/
/*  queue the MOUSEX and MOUSEY devices, and tie them to	*/
/*  the LEFT MOUSE						*/
    tie (LEFTMOUSE, MOUSEX, MOUSEY);  
    qdevice(REDRAW);	/* queue the REDRAW device		*/

    mapcolor (8, 240, 240, 240);/*  make color of arcs, lines	*/
    mapcolor (9, 0, 175, 0);	/*  make color of grass		*/
    mapcolor (16, 240, 150, 0);  /*  make overlay (ball) 	*/
    mapcolor (24, 240, 150, 0);  /*  for colors 16, 24, and 25	*/
    mapcolor (25, 240, 150, 0);

/*  Now we define two pop-up menus using two different methods. The 1st */
/*  method defines the menu and the text describing the options, all in */
/*  the same step, using a defpup () call. The 2nd method first defines */
/*  the menu using a newpup () call, then puts in the text using a      */
/*  addtopup () all.                                                    */

/*  Note: the '|' character separates menu items, and the %t flag is    */
/*  used to designate the menu title.                                   */

/*  newpup () and defpup () both return the menu id number used later.  */

    killmenu = defpup ("Do you want to exit? %t|yes|no");
    menu = newpup ();
    addtopup (menu, "Baseball %t|exit program");

}

/*  In processinput(), the RIGHTMOUSE button, when it is pressed	*/
/*  down (val == 1), displays a popup menu.  The value returned		*/
/*  by the user's selection is stored in variable menuval.  The		*/
/*  value returned may be:						*/
/*	 1 -- for exit program						*/
/*	-1 -- nothing on menu is chosen					*/
/*  Then the user is asked to reconfirm an exit program decision.	*/
/*	 1 -- for yes							*/
/*	 2 -- for no							*/
/*	-1 -- nothing on menu is chosen					*/

processinput() {
    short val;
    short mx, my;
    int dev;
    int domove;
    int menuval, kmenuval;

    domove = FALSE;
    while (qtest()) {
	dev = qread(&val);
	switch (dev) { 
	case LEFTMOUSE:
	    if (val == 1) /*  when mouse pressed, will do motion  */
		domove = TRUE;	
	    break;
	case MOUSEX:
	    mx = val;
	    break;
	case MOUSEY:
	    my = val;
	    break;
	case RIGHTMOUSE:  /*  when RIGHT MOUSE pressed, do popup  */
	    if (val == 1) {
		menuval = dopup(menu); /* returns the menu item chosen */
		if (menuval == 1) {  /*  if exit chosen, ask user */
		    kmenuval = dopup (killmenu); /*  to reconfirm */
		    if (kmenuval == 1) 
			exit (0);
		}
	    }
	    break;
	case REDRAW:
	    reshapeviewport();
	    drawimage();
	    break;
	default:
	    break;
	}
    }
    if (domove)
	moveball(mx, my);
}

/*  drawimage() has not changed since the last example.		*/

drawimage () {
    frontbuffer(TRUE);	/*  draw the field twice, once to each buffer	*/
    color(BLACK);
    clear();
    diamond();
    frontbuffer(FALSE);
}

/*  moveball() has not changed since the last example.		*/

moveball(mx, my) 
short mx, my;
{
    int i;
    int screenx, screeny;	/*  position of lower left corner	*/
    int distx, disty;		/*  distance ball will travel		*/
    float incrx, incry;		/*  amount to move ball in 1 iteration	*/
    float newx, newy;		/*  location for ball in flight		*/
    int inewx, inewy;		/*  integer location for ball in flight	*/

    getorigin (&screenx, &screeny);
    distx = mx - screenx;
    disty = my - screeny;
    incrx = (float) distx / 50.0;
    incry = (float) disty / 50.0;
    newx = 0.0;
    newy = 0.0;
    writemask(16);	/*  protect the first four bitplanes		*/
    for (i = 0; i < 50; i = i + 1) {
	newx = newx + incrx;
	newy = newy + incry;
	inewx = (int) newx;
	inewy = (int) newy;
	color (BLACK);		/*  clear away old ball			*/
	clear ();
	color (16);
	circfs (inewx, inewy, 3);/*  draw ball at position (inewx, inewy)*/
	swapbuffers();
    }
	/*  Draw the ball into both front and back buffers.	*/
    frontbuffer (TRUE);	/*  draw final resting place of ball to		*/
    color (BLACK);	/*  both buffers				*/
    clear ();
    color (16);
    circfs (inewx, inewy, 3);/*  draw ball at position (inewx, inewy)*/
    frontbuffer(FALSE);
    writemask(0xfff);	/*  unprotect all bitplanes		*/
}

/*  diamond() has not changed since the last example.		*/

diamond () {
    color (9);		/*  make color 9 current color	*/
    arcfs (0, 0, 375, 0, 900);	/*  grass		*/
    color (8);		/*  make color 8 current color	*/
    linewidth(2);	/*  change thickness of lines	*/
    arcs (0, 0, 375, 0, 900);	/*  fences		*/
    arcs (0, 0, 150, 0, 900);	/*  infield		*/

    move2s (0, 0);		/*  foul lines		*/
    draw2s (0, 400);
    move2s (0, 0);
    draw2s (400, 0);
    linewidth(1);	/*  restore thickness of lines	*/

    circfs (43, 43, 10);	/*  pitcher's mound	*/

    draw_base (90, 0);		/*  first, second and third bases	*/
    draw_base (90, 90);
    draw_base (0, 90);

    pmv2s (0, 0);		/*  draw home plate	*/
    pdr2s (0, 3);
    pdr2s (3, 6);
    pdr2s (6, 3);
    pdr2s (3, 0);
    pclos ();

    cmov2i (100, 400);	/*  position to draw scoreboard	*/
    charstr ("New York 3");
    cmov2i (100, 385);	
    charstr ("Boston   2");
}

/*  draw_base has not changed since the last example.			*/

draw_base (x, y)
short   x,
        y;
{
    pmv2s (x, y);
    rpdr2s (5, 0);
    rpdr2s (0, 5);
    rpdr2s (-5, 0);
    pclos ();
}

File: gifts.gl2/fgl_tutor/src/workshop/menu8.f
C  menu8.f -- This program allows the user a pop-up menu	
C  interface to an application program.  In this example,	
C  the RIGHT MOUSE button controls a popup menu.  This popup	
C  menu can be used to exit the program.  Intention to exit	
C  the program must be reconfirmed by the user responding to	
C  a second popup menu.  Popup menu support is ONLY available	
C  in the window manager.					
C  graphics library calls introduced:  defpup, newpup,	
C		addtop, dopup.				

#include "fgl.h"
#include "fdevice.h"


C  new global variables for menus		

	INTEGER  MENU, KILLMENU
	COMMON /MENB/MENU,KILLMENU

C  the main routine has not changed since the last example.	

	CALL INITIALIZE
	CALL DRAWIMAGE

 100   IF (QTEST() .EQ. 0) THEN
		CALL SWAPBU
	ENDIF
	CALL PROCESSINPUT
	GOTO 100

	END

C  In initialize, queue the RIGHTMOUSE button for popup menus.	
C  Also the pup calls are used to create popup menus.			
	SUBROUTINE INITIALIZE

#include "fgl.h"
#include "fdevice.h"

	INTEGER DUMMY
	INTEGER  MENU, KILLMENU
	COMMON /MENB/MENU,KILLMENU

	CALL PREFSI(450, 450)
	DUMMY = WINOPE ('DIAMOND', 7)

C   display mode to become double buffer
	CALL DOUBLE
C   display mode change takes effect
	CALL GCONFI

	CALL QDEVIC(LEFTMO)
	CALL QDEVIC(RIGHTM)
	CALL TIE (LEFTMO, MOUSEX, MOUSEY)
	CALL QDEVIC( REDRAW )

C   make color of fences, lines
	CALL MAPCOL (8, 240, 240, 240)

C   make color of grass
	CALL MAPCOL (9, 0, 175, 0)

C   make overlay (ball) for colors 16, 24, 25
	CALL MAPCOL (16, 240, 150, 0)
	CALL MAPCOL (24, 240, 150, 0)
	CALL MAPCOL (25, 240, 150, 0)

C  Now we define two pop-up menus using newpup call, then we put in the 
C  text using an addtop all.							   

C  Note: the '|' character separates menu items, and the %t flag is    
C  used to designate the menu title.					

C  newpup returns the menu id number used later.  

	KILLMENU = NEWPUP()
	CALL ADDTOP(KILLMENU, 'Do you want to exit? %t|yes|no', 30, 1)
	MENU = NEWPUP ()
	CALL ADDTOP (MENU, 'Baseball %t|exit program', 24, 1)

	RETURN
	END

C  In processinput, the RIGHTM button, when it is pressed	
C  down (val .eq. 1), displays a popup menu.  The value returned	
C  by the user's selection is stored in variable menuval.  The		
C  value returned may be:						
C	 1 -- for exit program						
C	-1 -- nothing on menu is chosen					
C  Then the user is asked to reconfirm an exit program decision.	
C	 1 -- for yes							
C	 2 -- for no							
C	-1 -- nothing on menu is chosen					

	SUBROUTINE PROCESSINPUT

#include "fgl.h"
#include "fdevice.h"

	INTEGER*2 VAL
	INTEGER*2 MX, MY
	INTEGER DEV
	LOGICAL DOMOVE
	INTEGER*2 MENUVAL, KMENUVAL
	INTEGER  MENU, KILLMENU
	COMMON /MENB/MENU,KILLMENU

	DOMOVE = .FALSE.
 300   IF (QTEST() .EQ. 0) GOTO 350
	     DEV = QREAD(VAL)
	     IF (DEV .EQ. LEFTMO) THEN
C		When the mouse is pressed down, will do motion
		IF (VAL .EQ. 1) THEN
		    DOMOVE = .TRUE.
		ENDIF
	    ELSE IF (DEV .EQ. MOUSEX) THEN
		MX = VAL
	    ELSE IF (DEV .EQ. MOUSEY) THEN
		MY = VAL
	    ELSE IF (DEV .EQ. RIGHTM) THEN
C	    When RIGHTM is pressed, do the popup.
		IF (VAL .EQ. 1) THEN
		    MENUVAL = DOPUP(MENU)
C		    If exit, then get confirmation.
		    IF (MENUVAL .EQ. 1) THEN
			KMENUVAL = DOPUP(KILLMENU)
			IF (KMENUVAL .EQ. 1) THEN
			    STOP
			ENDIF
		    ENDIF
		ENDIF
	    ELSE IF (DEV .EQ. REDRAW) THEN
		CALL RESHAP
		CALL DRAWIMAGE
	    ENDIF
	    GOTO 300
 350   CONTINUE

	IF (DOMOVE) THEN
	    CALL MOVEBALL(MX, MY)
	ENDIF

	RETURN
	END


C   drawimage has not changed since the last example

	SUBROUTINE DRAWIMAGE

#include "fgl.h"
#include "fdevice.h"

C   draw the field twice, once into each buffer
	CALL FRONTB(.TRUE.)
	CALL COLOR(0)
	CALL CLEAR
	CALL DIAMOND
	CALL FRONTB(.FALSE.)

	RETURN
	END

C  moveball() has not changed since the last example.		

	SUBROUTINE MOVEBALL(MX, MY)

	INTEGER*2 MX, MY

#include "fgl.h"
#include "fdevice.h"

	INTEGER I

C   screenx, screeny are position of lower left corner
C   distx, disty are distance ball will travel
C   incrx, incry are amount to move ball in 1 iteration
C   location of ball in flight
C   integer location of ball in flight

	INTEGER SCREENX, SCREENY
	INTEGER DISTX, DISTY
	REAL INCRX, INCRY
	REAL NEWX, NEWY
	INTEGER INEWX, INEWY

	CALL GETORI(SCREENX, SCREENY)
	DISTX = MX - SCREENX
	DISTY = MY - SCREENY
	INCRX = DISTX/50.0
	INCRY = DISTY/50.0
	NEWX = 0.0
	NEWY = 0.0

C   protect the first four bitplanes
	CALL WRITEM(16)

	DO 400 I=0, 49, 1

	     NEWX = NEWX + INCRX
	     NEWY = NEWY + INCRY
	     INEWX = NEWX
	     INEWY = NEWY

C	  clear away old ball
	     CALL COLOR(0)
	     CALL CLEAR

C	  draw ball at position (x, y)
	     CALL COLOR(16)
	     CALL CIRCFI(INEWX, INEWY, 3)

C	  swapbuffers
	     CALL SWAPBU

 400   CONTINUE

C   Draw the ball into both front and back buffers.
	CALL FRONTB(.TRUE.)
	CALL COLOR(0)
	CALL CLEAR
	CALL COLOR(16)
	CALL CIRCFI(INEWX, INEWY, 3)
	CALL FRONTB(.FALSE.)

C   Unprotect all bitplanes.
	CALL WRITEM($FFF)

	RETURN
	END

C   diamond has not changed since last example.

	SUBROUTINE DIAMOND

#include "fgl.h"
#include "fdevice.h"

C   draw grass
	CALL COLOR(9)
	CALL ARCFI (0, 0, 375, 0, 900)

	CALL COLOR(8)

C   change thickness of lines
	CALL LINEWI(2)

C   fences
	CALL ARCI(0, 0, 375, 0, 900)
	CALL ARCI(0, 0, 150, 0, 900)

C   foul lines
	CALL MOVE2I(0, 0)
	CALL DRAW2I(0, 400)
	CALL MOVE2I(0, 0)
	CALL DRAW2I(400, 0)

C   restore thickness of lines
	CALL LINEWI(1)

C   pitcher's mound
	CALL CIRCFI(43, 43, 10)

C   first, second and third base
	CALL DRAWBASE(90, 0)
	CALL DRAWBASE(90, 90)
	CALL DRAWBASE(0, 90)

C   draw home plate
	CALL PMV2I(0, 0)
	CALL PDR2I(0, 3)
	CALL PDR2I(3, 6)
	CALL PDR2I(6, 3)
	CALL PDR2I(3, 0)
	CALL PCLOS

C   draw scoreboard
	CALL CMOV2I (100, 400)
	CALL CHARST ('New York 3', 10)
	CALL CMOV2I (100, 385)
	CALL CHARST ('Boston   2', 10)

	RETURN
	END

C  draw_base has not changed since the last example.

	SUBROUTINE DRAWBASE(X, Y)

	INTEGER*2 X,Y

#include "fgl.h"
#include "fdevice.h"

	CALL PMV2I(X, Y)
	CALL RPDR2I(5, 0)
	CALL RPDR2I(0, 5)
	CALL RPDR2I(-5, 0)
	CALL PCLOS

	RETURN
	END

